crates/goose/src/agents/truncate.rs

```
@@ -246,152 +101,171 @@ impl Agent for TruncateAgent {
+                            .filter_map(|content| content.as_tool_request())
+                            .collect();
+
+                        if tool_requests.is_empty() {
+                            break;
+                        }
+
+                        // Then dispatch each in parallel
+                        let futures: Vec<_> = tool_requests
+                            .iter()
+                            .filter_map(|request| request.tool_call.clone().ok())
+                            .map(|tool_call| capabilities.dispatch_tool_call(tool_call))
+                            .collect();
+
+                        // Process all the futures in parallel but wait until all are finished
+                        let outputs = futures::future::join_all(futures).await;
+
+                        // Create a message with the responses
+                        let mut message_tool_response = Message::user();
+                        // Now combine these into MessageContent::ToolResponse using the original ID
+                        for (request, output) in tool_requests.iter().zip(outputs.into_iter()) {
+                            message_tool_response = message_tool_response.with_tool_response(
+                                request.id.clone(),
+                                output,
+                            );
+                        }
+
+                        yield message_tool_response.clone();
+
+                        messages.push(response);
+                        messages.push(message_tool_response);
+                    },
+                    Err(ProviderError::ContextLengthExceeded(_)) => {
+                        if truncation_attempt >= MAX_TRUNCATION_ATTEMPTS {
+                            // Create a user-facing error message & terminate the stream
+                            let error_message = Message::user().with_text("Error: Context length exceeds limits even after multiple attempts to truncate.");
```

@baxen - "i think the previous message would have been a user message (e.g. before any tool calls, this is just after the input message. at the start of a loop after a tool call, it would be after a tool_use assistant followed by a tool_result user). so we need to surface an assistant error here instead if we treat it as a message?"